sympolyc/src/expr.rs

```diff
@@ -0,0 +1,241 @@
+    fn default() -> Self {
+        MVPoly::new()
+    }
+}
+
+// Addition
+impl<F: PrimeField, const N: usize, const D: usize> Add for MVPoly<F, N, D> {
+    type Output = Self;
+
+    fn add(self, other: Self) -> Self {
+        let mut result = MVPoly::new();
+        for i in 0..self.coeff.len() {
+            result.coeff[i] = self.coeff[i] + other.coeff[i];
+        }
+        result
+    }
+}
+
+impl<F: PrimeField, const N: usize, const D: usize> Add<&MVPoly<F, N, D>> for MVPoly<F, N, D> {
+    type Output = MVPoly<F, N, D>;
+
+    fn add(self, other: &MVPoly<F, N, D>) -> MVPoly<F, N, D> {
+        let mut result = MVPoly::new();
+        for i in 0..self.coeff.len() {
+            result.coeff[i] = self.coeff[i] + other.coeff[i];
+        }
+        result
+    }
+}
+
+impl<F: PrimeField, const N: usize, const D: usize> Add<MVPoly<F, N, D>> for &MVPoly<F, N, D> {
+    type Output = MVPoly<F, N, D>;
+
+    fn add(self, other: MVPoly<F, N, D>) -> MVPoly<F, N, D> {
+        let mut result = MVPoly::new();
+        for i in 0..self.coeff.len() {
+            result.coeff[i] = self.coeff[i] + other.coeff[i];
+        }
+        result
+    }
+}
+
+impl<F: PrimeField, const N: usize, const D: usize> Add<&MVPoly<F, N, D>> for &MVPoly<F, N, D> {
+    type Output = MVPoly<F, N, D>;
+
+    fn add(self, other: &MVPoly<F, N, D>) -> MVPoly<F, N, D> {
+        let mut result = MVPoly::new();
+        for i in 0..self.coeff.len() {
+            result.coeff[i] = self.coeff[i] + other.coeff[i];
+        }
+        result
+    }
+}
+
+// Multiplication
+impl<F: PrimeField, const N: usize, const D: usize> Mul<MVPoly<F, N, D>> for MVPoly<F, N, D> {
+    type Output = Self;
+
+    fn mul(self, other: Self) -> Self {
+        let mut acc = HashMap::new();
+        let mut prime_gen = PrimeNumberGenerator::new();
+        let mut result = vec![];
+        (0..self.coeff.len()).for_each(|i| {
+            let mut sum = F::zero();
+            let normalized_index = self.normalized_indices[i];
+            let two_factors_decomposition =
+                compute_all_two_factors_decomposition(normalized_index, &mut acc, &mut prime_gen);
+            two_factors_decomposition.iter().for_each(|(a, b)| {
+                // FIXME: we should keep the inverse normalized indices
+                let inv_a = self
+                    .normalized_indices
+                    .iter()
+                    .position(|&x| x == *a)
+                    .unwrap();
+                let inv_b = self
+                    .normalized_indices
+                    .iter()
+                    .position(|&x| x == *b)
+                    .unwrap();
+                let a_coeff = self.coeff[inv_a];
+                let b_coeff = other.coeff[inv_b];
+                let product = a_coeff * b_coeff;
+                sum += product;
+            });
+            result.push(sum);
+        });
+        Self::from_coeffs(result)
```

Comment:
Here the normalized indices will be computed in function of the degree bound D, which is not invariant by multiplication.
Can't that be an issue ?

Reply:
I forgot to mention in the doc: it is expected for now. This first PR is a PoC regarding multiplication using prime numbers. I will change later or add some warnings/raise exception in future PR.

Reply:
Note that it is like this for now as it is initially supposed to be used from the expression framework, and with constraints where we know the number of variables and the degree.